src/test/groovy/org/prebid/server/functional/tests/BidAdjustmentSpec.groovy

```diff
@@ -0,0 +1,136 @@
+package org.prebid.server.functional.tests
+
+import org.prebid.server.functional.model.request.auction.BidAdjustmentFactors
+import org.prebid.server.functional.model.request.auction.BidRequest
+import org.prebid.server.functional.model.response.auction.BidResponse
+import org.prebid.server.functional.service.PrebidServerException
+import org.prebid.server.functional.util.PBSUtils
+
+import static io.netty.handler.codec.http.HttpResponseStatus.BAD_REQUEST
+import static org.prebid.server.functional.model.bidder.BidderName.APPNEXUS
+import static org.prebid.server.functional.model.bidder.BidderName.GENERIC
+import static org.prebid.server.functional.model.bidder.BidderName.RUBICON
+import static org.prebid.server.functional.model.request.auction.BidAdjustmentMediaType.BANNER
+import static org.prebid.server.functional.model.request.auction.BidAdjustmentMediaType.NATIVE
+import static org.prebid.server.functional.model.request.auction.BidAdjustmentMediaType.VIDEO
+import static org.prebid.server.functional.model.request.auction.DistributionChannel.SITE
+
+class BidAdjustmentSpec extends BaseSpec {
+
+    def "PBS should adjust bid price for matching bidder when request has per-bidder bid adjustment factors"() {
+        given: "Default bid request with bid adjustment"
+        def bidRequest = BidRequest.getDefaultBidRequest(SITE).tap {
+            ext.prebid.bidAdjustmentFactors = new BidAdjustmentFactors(adjustments: [(GENERIC): bidAdjustmentFactor])
+        }
+
+        and: "Default bid response"
+        def bidResponse = BidResponse.getDefaultBidResponse(bidRequest)
+        bidder.setResponse(bidRequest.id, bidResponse)
+
+        when: "PBS processes auction request"
+        def response = defaultPbsService.sendAuctionRequest(bidRequest)
+
+        then: "Final bid price should be adjusted"
+        assert response?.seatbid[0]?.bid[0]?.price == bidResponse.seatbid[0].bid[0].price * bidAdjustmentFactor
```

Comment:
If seatbid is null, so then seatbid[0] will throw a NPE. Groovy null-safe operator can't help when you access an element with an array index operator. To guarantee a null-safety '?.' operator should be used as seatbid?.first().
Same for other places

Reply:
Derp. Good catch.